src/Build/Xml/XmlReaderExtension.cs

```diff
@@ -34,12 +37,22 @@ private XmlReaderExtension(string file)
                 _stream = new FileStream(file, FileMode.Open, FileAccess.Read, FileShare.Read);
                 _streamReader = new StreamReader(_stream, Encoding.UTF8, true);
                 Reader = GetXmlReader(_streamReader, out _encoding);
+                var bom = EncodingUtilities.StreamStartsWithBom(_stream);
 
                 // Override detected encoding if xml encoding attribute is specified
                 var encodingAttribute = Reader.GetAttribute("encoding");
                 _encoding = !string.IsNullOrEmpty(encodingAttribute)
                     ? Encoding.GetEncoding(encodingAttribute)
                     : _encoding;
+
+                // If the encoding is UTF8 but there was not a BOM detected in the file then we need
+                // to override the detected Encoding with UTF8 and encoderShouldEmitUTF8Identifier
+                // set to false. This is to prevent the BOM from being added if we save the file
+                // later using the detected encoding.
+                if (Equals(_encoding, Encoding.UTF8) && !bom)
```

Comment:
We talked offline and figured out why `StreamReader` couldn't be trusted in the case - the parameter that detects encoding is called `detectEncodingFromByteOrderMarks` _not_ `detectEncoding`. ie It will only change the "default" encoding passed to the constructor if the byte-order-mark is present, not if it isn't.

If Andy changed the call to StreamReader to pass s_utf8NoBom by default, then this will do the right thing and he can end up reading the Encoding from the StreamReader instead of checking the preamble himself.